arch/arm/core/swap.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2018 Linaro, Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <toolchain.h>
+#include <kernel_structs.h>
+
+#ifdef CONFIG_EXECUTION_BENCHMARKING
+extern void read_timer_start_of_swap(void);
+#endif
+extern const int _k_neg_eagain;
+
+/**
+ *
+ * @brief Initiate a cooperative context switch
+ *
+ * The __swap() routine is invoked by various kernel services to effect
+ * a cooperative context context switch.  Prior to invoking __swap(), the caller
+ * disables interrupts via irq_lock() and the return 'key' is passed as a
+ * parameter to __swap().  The 'key' actually represents the BASEPRI register
+ * prior to disabling interrupts via the BASEPRI mechanism.
+ *
+ * __swap() itself does not do much.
+ *
+ * It simply stores the intlock key (the BASEPRI value) parameter into
+ * current->basepri, and then triggers a service call exception (svc) to setup
+ * the PendSV exception, which does the heavy lifting of context switching.
+
+ * This is the only place we have to save BASEPRI since the other paths to
+ * __pendsv all come from handling an interrupt, which means we know the
+ * interrupts were not locked: in that case the BASEPRI value is 0.
+ *
+ * Given that __swap() is called to effect a cooperative context switch,
+ * only the caller-saved integer registers need to be saved in the thread of the
+ * outgoing thread. This is all performed by the hardware, which stores it in
+ * its exception stack frame, created when handling the svc exception.
+ *
+ * On Cortex-M0/M0+ the intlock key is represented by the PRIMASK register,
+ * as BASEPRI is not available.
+ *
+ * @return may contain a return value setup by a call to
+ * _set_thread_return_value()
+ *
+ * C function prototype:
+ *
+ * unsigned int __swap (unsigned int basepri);
+ *
+ */
+int __swap(int key)
+{
+#ifdef CONFIG_USERSPACE
+	/* Save off current privilege mode */
+	u32_t mode;
+
+	__asm__ volatile("mrs %0, CONTROL\n\t" : "=r"(mode));
+	mode &= 0x3;
+	_current->arch.mode = mode;
+#endif
+#ifdef CONFIG_EXECUTION_BENCHMARKING
+	read_timer_start_of_swap();
+#endif
+
+	/* store off key and return value */
+	_current->arch.basepri = key;
+	_current->arch.swap_return_value = _k_neg_eagain;
+
+	/* set pending bit to make sure we */
+	SCB->ICSR |= SCB_ICSR_PENDSVSET_Msk;
```

Comment:
[That's just a request for clarification:]

So, as far as I understand, for ARMv6-M implementation there is no change.
For ARMv7-M we replace "svc #0" with the "setting" of the PendSV bit.
 Are we certain that this is sufficient? I mean, don't we need any changes on the PendSV handling (or even the SVC handling), in the wake of this change?

Reply:
Yes, for v6m, it is identical.  For v7m+, this just removes the svc #0 call to do the actual setting of the pendsv bit.  So there are a few scenarios we could be in when we are context switching:
1) We are currently in handler mode, handling an irq
2) We are currently in thread mode, with no irq's pending
3) We are currently in thread mode, with pending irqs of some sort.

The _Swap() function does an irq_lock() before it ever calls this function.  So we could still get interrupted, but only by things higher priority than the normal IRQS (fault handlers).  So if we have pending IRQs they are of normal or lower priority.

For the first case, we'll set the pendsv bit and it'll context switch at the tail end of irq handling (due to pendsv being 0xff prio).  No change in behavior here.

For the second case, we'll take the pendsv irq immediately after unlocking irqs.  No change here.  The only question is the latency between enabling irqs and getting the pendsv.  I'd expect it to be immediate.

For the third case, we'll set the pendsv bit, unlock irqs, and immediately take the other irqs, followed by the pendsv on the tail end.  No change in behavior here.

In all cases, I see equivalent behavior.

Reply:
@agross-linaro  thanks for clarifying this